File: pkg/wire/frame.go
// Package wire defines Sextant's wire atom: a JSON frame the bus stamps around a
// typed lexicon record. The record is user space (the client supplies it); the
// frame is bus space (the bus produces it). The frozen wrapper core
// (ID/Author/Kind/Epoch) is the only thing a protocol-epoch bump protects;
// Record is the free-evolution zone.
//
// See ADR-0006 (the wire atom), ADR-0010 (lifecycle & versioning), and ADR-0018
// / ADR-0019 (the bus implements the protocol and stamps the frame).
package wire

import (
	"encoding/json"
	"time"

	"github.com/oklog/ulid/v2"
)

// Frame kinds. Kind discriminates a frame: a message in flight or an artifact at
// rest (frame.json knownValues).
const (
	KindMessage  = "message"
	KindArtifact = "artifact"
)

// Epoch is the protocol epoch the SDK writes and checks. It bumps only on a
// breaking wire change; additive changes never bump it. See ADR-0010.
const Epoch = 1

// SkewTolerance is the default maximum allowed difference between a ULID's
// embedded timestamp and the bus-stamped time. See ADR-0006.
const SkewTolerance = 5 * time.Minute

// Lexicon is the typed content a frame carries: an AT-Protocol lexicon record,
// JSON by convention for now. It is an alias today, which names the concept and
// leaves a seam to attach validation or codegen later without changing the
// Frame.Record type here.
type Lexicon = json.RawMessage

// Frame is the wire atom: the bus-stamped wrapper around a typed record. The
// record is user space (the client supplies it); the frame is bus space (the bus
// produces id, author, and the rest). Kind discriminates it: a message in flight
// or an artifact at rest. See ADR-0018/ADR-0019.
type Frame struct {
	// ID is a ULID; it is also the dedup key. Bus-stamped (ADR-0019). For a
	// message, its embedded timestamp is not a trusted clock (see CheckSkew).
	ID string `json:"id"`
	// Author is the authenticated identity the bus took from the request. Not
	// client-set, and not forgeable by editing the record (ADR-0019).
	Author string `json:"author"`
	// Kind is the frame kind: KindMessage or KindArtifact.
	Kind string `json:"kind"`
	// Epoch is the protocol epoch this frame was written under, checked
	// per-message because durable streams outlive epochs.
	Epoch int `json:"epoch"`
	// Record is the typed content: an AT-Protocol lexicon record, opaque to the
	// bus and always JSON on this wire. Binary is a first-class bytes/blob value
	// within it — inline as {"$bytes": base64} now, native under a later
	// DAG-CBOR encoding, or a blob reference when large (ADR-0016).
	Record Lexicon `json:"record"`

	// The following are bus-stamped and present only when Kind is KindArtifact.

	// Revision is the artifact revision; it advances on each write. Bus-stamped.
	Revision uint64 `json:"revision,omitempty"`
	// CreatedAt is the artifact's creation time (RFC3339). Bus-stamped.
	CreatedAt string `json:"createdAt,omitempty"`
	// UpdatedAt is the time of the latest artifact write (RFC3339). Bus-stamped.
	UpdatedAt string `json:"updatedAt,omitempty"`
}

// New builds a valid message frame for the current epoch: a fresh ULID stamped
// at the current time, KindMessage, and the given author and record. Under
// ADR-0019 the bus owns frame stamping; this is the message constructor used
// until the bus serves the publish operation.
func New(author string, record Lexicon) Frame {
	return Frame{
		ID:     ulid.Make().String(),
		Author: author,
		Kind:   KindMessage,
		Epoch:  Epoch,
		Record: record,
	}
}

// Encode marshals the frame to its JSON wire form.
func Encode(f Frame) ([]byte, error) {
	return json.Marshal(f)
}

// Decode unmarshals a frame from its JSON wire form.
func Decode(b []byte) (Frame, error) {
	var f Frame
	err := json.Unmarshal(b, &f)
	return f, err
}

File: clients/ts/sdk/src/wire/frame.ts
// The wire atom: the bus-stamped wrapper around a typed lexicon record. The
// record is user space (the client supplies it); the frame is bus space (the
// bus produces id, author, and the rest). Mirrors protocol/wire/frame.go and
// protocol/wire/validate.go — the structure and the validation are the protocol
// contract every co-equal client reproduces (ADR-0006, ADR-0041).

import type { JSONValue } from "../types.js";

// Frame kinds. Kind discriminates a frame: a message in flight or an artifact
// at rest.
export const KIND_MESSAGE = "message";
export const KIND_ARTIFACT = "artifact";

// Frame is the wire atom — exactly the wire keys. A MESSAGE frame carries the
// five core keys (author, epoch, id, kind, record); the artifact-only keys
// (revision, createdAt, updatedAt) are OMITTED when absent, matching Go's
// `omitempty`. The codec never emits them on a message.
export interface Frame {
  id: string; // bus-minted ULID (26-char Crockford base32)
  author: string; // authenticated client id
  kind: string; // "message" | "artifact"
  epoch: number; // must equal EPOCH = 1
  record: JSONValue; // opaque lexicon; non-empty, valid JSON
  revision?: number; // artifact-only; omit on messages
  createdAt?: string; // artifact-only RFC3339; omit on messages
  updatedAt?: string; // artifact-only RFC3339; omit on messages
}

// Crockford base32 alphabet (excludes I, L, O, U), as used by ULID. The
// timestamp portion of a valid ULID can encode at most 0x7FFFFFFFFFFF ms (48
// bits), so the first character must be 0–7. Mirrors the parse oklog/ulid does.
const CROCKFORD = "0123456789ABCDEFGHJKMNPQRSTVWXYZ";
const DECODE: Record<string, number> = {};
for (let i = 0; i < CROCKFORD.length; i++) {
  const c = CROCKFORD[i]!;
  DECODE[c] = i;
  DECODE[c.toLowerCase()] = i;
}

// parseULIDMillis validates a 26-char Crockford-base32 ULID and returns the
// embedded millisecond timestamp (the high 48 bits). It throws on a malformed
// id, mirroring the strict parse the Go SDK relies on (ulid.Parse).
export function parseULIDMillis(id: string): number {
  if (id.length !== 26) {
    throw new Error(`wire: invalid ULID ${JSON.stringify(id)}: length ${id.length}, want 26`);
  }
  for (let i = 0; i < 26; i++) {
    if (DECODE[id[i]!] === undefined) {
      throw new Error(`wire: invalid ULID ${JSON.stringify(id)}: bad character at ${i}`);
    }
  }
  // The first character holds the top 3 bits of the 48-bit timestamp; a valid
  // ULID timestamp fits 48 bits, so it must be 0–7 (anything larger overflows).
  if (DECODE[id[0]!]! > 7) {
    throw new Error(`wire: invalid ULID ${JSON.stringify(id)}: timestamp overflow`);
  }
  // Decode the first 10 characters (50 bits, of which the top 48 are the
  // timestamp) into a millisecond value. 48 bits fits in a JS safe integer.
  let ms = 0;
  for (let i = 0; i < 10; i++) {
    ms = ms * 32 + DECODE[id[i]!]!;
  }
  return ms;
}

// isValidULID reports whether id parses as a ULID, without throwing.
export function isValidULID(id: string): boolean {
  try {
    parseULIDMillis(id);
    return true;
  } catch {
    return false;
  }
}

// validateFrame checks a frame is structurally well-formed — a parseable ULID
// id, a non-empty author, a known kind, and a non-empty record — mirroring
// wire/validate.go's Frame.Validate. It does NOT check the epoch: that is
// contextual (see checkEpoch), because durable streams outlive epochs. Throws a
// descriptive error on the first violation.
export function validateFrame(f: Frame): void {
  if (!isValidULID(f.id)) {
    throw new Error(`wire: invalid id: ${JSON.stringify(f.id)}`);
  }
  if (f.author === "") {
    throw new Error("wire: empty author");
  }
  if (f.kind !== KIND_MESSAGE && f.kind !== KIND_ARTIFACT) {
    throw new Error(`wire: unknown kind ${JSON.stringify(f.kind)}`);
  }
  // A present record. Go's check is `len(record) == 0 || !json.Valid(record)`
  // on the raw bytes — i.e. the record must be present and parse as JSON. After
  // decoding, any value other than an absent field is a valid, non-empty JSON
  // value (even `{}`, `[]`, `null`, which Go accepts as valid non-zero bytes),
  // so the faithful TS equivalent is: the field must be present (not undefined).
  if (f.record === undefined) {
    throw new Error("wire: record must be non-empty valid JSON");
  }
}
